python/lsst/obs/base/gen3/ingest.py

```diff
@@ -0,0 +1,537 @@
+        dtype=str,
+        allowed={"ignore": ("Do not add the new file to the Collection.  If "
+                            "'stash' is not None, the new file will be "
+                            "ingested into the stash Collection instead."),
+                 "fail": ("Raise RuntimeError if a conflict is encountered "
+                          "(which may then be caught if onError == 'continue')."),
+                 },
+        optional=False,
+        default="ignore",
+    )
+    stash = Field(
+        "Name of an alternate Collection to hold Datasets that lose conflicts.",
+        dtype=str,
+        default=None,
+    )
+    onError = ChoiceField(
+        "What to do if an error (including fatal conflicts) occurs.",
+        dtype=str,
+        allowed={"continue": "Warn and continue with the next file.",
+                 "break": ("Stop processing immediately, but leave "
+                           "already-ingested datasets in the repository."),
+                 "rollback": ("Stop processing and attempt to remove aleady-"
+                              "ingested datasets from the repository."),
+                 },
+        optional=False,
+        default="continue",
+    )
+
+
+class RawIngestTask(Task, metaclass=ABCMeta):
+    """Driver Task for ingesting raw data into Gen3 Butler repositories.
+
+    This Task is intended to be runnable from the command-line, but it doesn't
+    meet the other requirements of CmdLineTask or PipelineTask, and wouldn't
+    gain much from being one.  It also wouldn't really be appropriate as a
+    subtask of a CmdLineTask or PipelineTask; it's a Task essentially just to
+    leverage the logging and configurability functionality that provides.
+
+    Each instance of `RawIngestTask` writes to the same Butler and maintains a
+    cache of DataUnit entries that have already been added to or extracted
+    from its Registry.  Each invocation of `RawIngestTask.run` ingests a list
+    of files (possibly semi-atomically; see `RawIngestConfig.onError`).
+
+    RawIngestTask should be subclassed to specialize ingest for the actual
+    structure of raw data files produced by a particular camera. Subclasses
+    must either provide populated `MetadataReader` instances in the
+    `dataIdReader`, `visitReader`, and `exposureReader` class attributes, or
+    alternate implementations of the `extractDataId`, `extractVisit`, and
+    `extractExposure` methods that do not use those attributes (each
+    attribute-method pair may be handled differently).  Subclasses may also
+    wish to override `getFormatter` and/or (rarely) `getDatasetType`.  We do
+    not anticipate overriding `run`, `ensureDataUnits`, `ingestFile`, or
+    `processFile` to ever be necessary.
+
+    Parameters
+    ----------
+    config : `RawIngestConfig`
+        Configuration for whether/how to transfer files and how to handle
+        conflicts and errors.
+    butler : `daf.butler.Butler`
```

Comment:
All your class references in the docstrings need the `lsst.` in front. Consider using `~lsst.daf.butler.Butler` etc.